Allow plugins to bring new Vm vendor_types

agrare opened the pull request:
Currently every provider plugin that refreshes a VM has to add a line to the core VmOrTemplate.VENDOR_TYPES constant array which is obviously not very pluggable.

Not requiring that all vendor types be in this list and allowing subclasses to override `#vendor_display` with their own constant we can allow plugins to work without modifying core

https://github.com/ManageIQ/manageiq/issues/19440

agrare commented on app/models/vm_or_template.rb:
```diff
@@ -47,11 +47,19 @@ class VmOrTemplate < ApplicationRecord
     "unknown"     => "Unknown"
```
Not all of these have equivalent provider plugins (e.g. parallels, xen) so we can't completely remove a small list in core

Fryguy replied:
Where do we even use those vendor types?

agrare replied:
They are set by the provider in refresh: https://github.com/ManageIQ/manageiq-providers-amazon/blob/master/app/models/manageiq/providers/amazon/inventory/parser/cloud_manager.rb#L57

And shown on the UI https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/helpers/textual_mixins/textual_vmm_info.rb#L27-L42

chessbyte replied:
@agrare but do we support Xen and Parallels, in particular?  I highly doubt it.  So, let's get rid of them, if possible.

agrare commented:
@Fryguy I updated this to remove the validate inclusion so that provider plugins can override the `#vendor_display` method without having to add anything to core and updated the provider generator to set this.

Fryguy commented:
I was thinking about this more and I think we have to keep vendor unfortunately, as automate code and API calls likely use this value and I'm concerned about breaking them.

agrare commented:
Yeah this isn't dropping the vendor column, but dropping the requirement that all vendors be in this core array for the purpose of `vendor_display`

Fryguy commented:
I guess my  concern is that without the validation, we can't ensure that a set of known values are brought (previously the PR had plugins bringing a known vendor _and_ vendor_display value - i.e. it forced them to pick a value for both)

agrare commented:
Okay I'll put the PR back the way it was originally then

agrare commented:
Hm using leaf_subclasses doesn't seem to be playing well constant autoloading

agrare commented:
Going to need to investigate a different approach to this
